File: commander.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "instrucao.h"

#define INIT_CODE "v3.0 hex words plain\n"
#define FILE_INPUT "input.asm"
#define FILE_OUTPUT "output.txt"

FILE *input;
FILE *output;

char init() {
    input = fopen(FILE_INPUT, "r");
    output = fopen(FILE_OUTPUT, "w");
    if (!input || !output) {
        printf("ERRO AO INICIAR ARQUIVOS.\n");
        return 0;
    }
    fprintf(output, INIT_CODE);
    return 1;
}

Instrucao lerComando(char linha[64]) {
    Instrucao in;
    memset(&in, 0, sizeof(Instrucao));
    sscanf(linha, "%5[A-Z] %3s %3s", in.comando, in.op1, in.op2);
    return in;
}

void salva(unsigned char dados[], int pos, int tam){
    for(int i = pos; i < tam; i++){
        dados[i] = 0x00;
    }

    for(int i = 0; i < tam; i++){
        fprintf(output, "%02X", dados[i]);
        if((i+1) % 16 == 0)
            fprintf(output, "\n");
        else
            fprintf(output, " ");
    }
    fclose(input);
    fclose(output);
}

File: assembler/commander.py
INIT_CODE = "v3.0 hex words plain\n"
FILE_INPUT = "input.asm"
FILE_OUTPUT = "output.txt"

class Instrucao:
    def __init__(self, comando="", op1="", op2=""):
        self.comando = comando
        self.op1 = op1
        self.op2 = op2

def init():
    try:
        input = open(FILE_INPUT, "r") 
        output = open(FILE_OUTPUT, "w")
        output.write(INIT_CODE)
        return input, output
    except IOError:
        print("ERRO AO INICIAR ARQUIVOS.")
        return None, None

def lerComando(linha):
    trecho = linha.split()
    comando = trecho[0]
    op1 = trecho[1]
    op2 = trecho[2]
    return Instrucao(comando, op1, op2)

def salva(dados, pos, tam, output_file):
    vet = dados.copy()
    for i in range(pos, tam):
        vet.append(0x00)

    for i in range(tam):
        output_file.write(f"{vet[i]:02x}")
        if (i+1) % 16 == 0:
            output_file.write("\n")
        else:
            output_file.write(" ")
    
    output_file.close()

inp, out = init()
x = lerComando("DATA R0 00")
ram = [0x20,0x00]
salva(ram,1,256, out)